### ISSUE TITLE
Pow doesn't infer types when variables are used

### ISSUE THREAD

**Issue by GatienBouyer** [python/mypy#16635]
**Bug Report**

I would expect that using variables to hold my values doesn't lose the resulting type.

Sample code:
```python
my_wanted_int = 2 ** 3
reveal_type(my_wanted_int) # = int

my_2 = 2
my_3 = 3
reveal_type(my_2) # = int
reveal_type(my_3) # = int
star_result2 = my_2 ** my_3
reveal_type(star_result2) # = Any
```

**Actual Behavior**

```
$ mypy -V
mypy 1.7.1 (compiled: yes)
$ mypy -c "
my_wanted_int = 2 ** 3
reveal_type(my_wanted_int) # = int

my_2 = 2
my_3 = 3
reveal_type(my_2) # = int
reveal_type(my_3) # = int
star_result2 = my_2 ** my_3
reveal_type(star_result2) # = Any instead of int
"
<string>:3: note: Revealed type is "builtins.int"
<string>:7: note: Revealed type is "builtins.int"
<string>:8: note: Revealed type is "builtins.int"
<string>:10: note: Revealed type is "Any"
Success: no issues found in 1 source file
```

Here is the example in the mypy playground:
https://mypy-play.net/?mypy=latest&python=3.11&gist=e53d4334e01415845108453985c8f88f

**Your Environment**

- Mypy version used: 1.7.1
- Python version used: 3.11

**Comment by JelleZijlstra** [comment]
@ikonst see https://github.com/python/typeshed/blob/ad5ec921e0c0da471867996fc85b94a3e7d83ecf/stdlib/builtins.pyi#L294 for why it's this way.

@GatienBouyer the explanation for what you're seeing is that if you put the value in a variable, mypy infers the type as just `int`, so it takes the `Any` path from above. But if you use the value directly, mypy knows the value is exactly 2 or 3, enabling it to use the overload for positive or negative integers that enables it to infer the value as `int`.

**Comment by ikonst** [comment]
Ahh, I see, e.g. existing code in https://github.com/Yelp/paasta/blob/master/paasta_tools/utils.py:
```py
def calculate_tail_lines(verbose_level: int) -> int:
    if verbose_level <= 1:
        return 0
    else:
        return 10 ** (verbose_level - 1)
```

Nobody says you can't call `calculate_tail_lines(0)` in which case the `int` type will be wrong, but I suppose they carefully pass `verbose_level > 0` and making that change would be a false positive for them.

**Comment by GatienBouyer** [comment]
As already discussed in https://github.com/python/typeshed/pull/7737#issuecomment-1253696117 and in https://github.com/python/typeshed/pull/6287#issuecomment-967377147, mypy has no way to distinguish between positive and negative integers. Until further change on this topic occurs, I close this issue.

**Comment by GatienBouyer** [comment]
I add this reference for completeness.
The original discussion: https://github.com/python/typeshed/pull/2271

**Comment by erictraut** [comment]
I'll note that pyright doesn't have a problem with the original code sample in this issue. That's because pyright [narrows to literal types on assignment](https://microsoft.github.io/pyright/#/mypy-comparison?id=assignment-based-narrowing-for-literals), whereas mypy does not. Perhaps this is something mypy maintainers would consider changing.

Code sample in [pyright playground](https://pyright-play.net/?code=LYTw%2Bg7ghgdgLgUwCZgJbwAQF4MCYBU%2BAzAFABOCAbglADZhwgAOCAFKJLIiunAJQYMAYgy8SJDrmx4J4ItNIVqdBszaSBwjO2YgAXKPgAaDExBlUAcwAWcAwBlUiMnQDauALp9yVGvUYsOmBEmiI6Zga8JmYWNnYYjs5uRF4kAM5wUGRgFGkArrRwUjiShByKvioBbBlZOQj5hbih2qARGACCMCDR5la2kfDeQA)

```python
my_wanted_int = 2**3
reveal_type(my_wanted_int)  # int

my_2 = 2
my_3 = 3
reveal_type(my_2)  # (mypy: int, pyright: Literal[2])
reveal_type(my_3)  # (mypy: int, pyright: Literal[3])
star_result2 = my_2**my_3
reveal_type(star_result2)  # (mypy: Any, pyright: int)
```
